pkg/aggregator/no_aggregation_stream_worker.go

```diff
@@ -0,0 +1,215 @@
+// really small payloads (that could have potentially been filled).
+type noAggregationStreamWorker struct {
+	serializer           serializer.MetricSerializer
+	flushConfig          FlushAndSerializeInParallel
+	maxMetricsPerPayload int
+
+	seriesSink   *metrics.IterableSeries
+	sketchesSink *metrics.IterableSketches
+
+	taggerBuffer *tagset.HashlessTagsAccumulator
+	metricBuffer *tagset.HashlessTagsAccumulator
+
+	samplesChan chan metrics.MetricSampleBatch
+	stopChan    chan trigger
+	flushChan   chan trigger
+}
+
+func newNoAggregationStreamWorker(maxMetricsPerPayload int, serializer serializer.MetricSerializer, flushConfig FlushAndSerializeInParallel) *noAggregationStreamWorker {
+	return &noAggregationStreamWorker{
+		serializer:           serializer,
+		flushConfig:          flushConfig,
+		maxMetricsPerPayload: maxMetricsPerPayload,
+
+		seriesSink:   nil,
+		sketchesSink: nil,
+
+		taggerBuffer: tagset.NewHashlessTagsAccumulator(),
+		metricBuffer: tagset.NewHashlessTagsAccumulator(),
+
+		stopChan:    make(chan trigger),
+		flushChan:   make(chan trigger),
+		samplesChan: make(chan metrics.MetricSampleBatch, config.Datadog.GetInt("dogstatsd_queue_size")),
+	}
+}
+
+func (w *noAggregationStreamWorker) addSamples(samples metrics.MetricSampleBatch) {
+	if len(samples) == 0 {
+		return
+	}
+	w.samplesChan <- samples
+}
+
+func (w *noAggregationStreamWorker) stop(wait bool) {
+	var blockChan chan struct{}
+	if wait {
+		blockChan = make(chan struct{})
+	}
+
+	trigger := trigger{
+		time:      time.Now(),
+		blockChan: blockChan,
+	}
+
+	w.stopChan <- trigger
+
+	if wait {
+		<-blockChan
+	}
+}
+
+func (w *noAggregationStreamWorker) flush(wait bool) {
+	var blockChan chan struct{}
+	if wait {
+		blockChan = make(chan struct{})
+	}
+
+	trigger := trigger{
+		time:      time.Now(),
+		blockChan: blockChan,
+	}
+
+	w.flushChan <- trigger
+
+	if wait {
+		<-blockChan
+	}
+}
+
+// mainloop of the no aggregation stream worker:
+//   * it receives samples and counts how much it has sent to the serializer, if it has more than a given amount it stops
+//     streaming for the serializer to start sending the payloads to the forwarder, and then starts the streaming
+//     mainloop again
+//   * it also checks every 2 seconds if it has stopped receiving samples, if so, it stops streaming to the
+//     the serializer for a while in order to let the serializer sends the payloads up to the forwarder, and starts
+//     the streaming mainloop again
+//   * listens for a stop signal
+//   * listens for a flush signal
+// This is not ideal since the serializer should automatically takes the decision when to flush payloads to
+// the serializer but that's not how it works today, see noAggregationStreamWorker comment.
+func (w *noAggregationStreamWorker) run() {
+	log.Debugf("Starting streaming routine for the no-aggregation pipeline")
+
+	ticker := time.NewTicker(time.Second * 2)
+	defer ticker.Stop()
+	lastStream := time.Now().Add(time.Hour * 24)
```

Comment:
What is the significance of this being 24 hours?

Reply:
A far enough date to not trigger an "automatic" flush too fast if one has just been done _and no metrics have been received meanwhile_. (that's the "and" which is important here)
So no significance, I'll add a comment on this 🙇